File: libs/party/src/system.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SystemPlayerEntry {
    pub rank: i64,
    pub player_id: String,
    pub name: String,
    pub accumulated_score: i64,
    pub crossword_score: Option<i64>,
    pub word_search_score: Option<i64>,
    pub quiz_score: Option<i64>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SystemAdminSnapshot {
    pub players: Vec<SystemPlayerEntry>,
}

#[derive(Debug, Clone)]
pub struct SystemScoreInput {
    pub player_id: String,
    pub name: String,
    pub connected: bool,
    pub crossword_score: Option<i64>,
    pub word_search_score: Option<i64>,
    pub quiz_score: Option<i64>,
}

fn accumulated_of(player: &SystemScoreInput) -> i64 {
    player.crossword_score.unwrap_or(0) + player.word_search_score.unwrap_or(0)
}

pub fn build_system_leaderboard(players: &[SystemScoreInput]) -> SystemAdminSnapshot {
    let mut ranked: Vec<&SystemScoreInput> = players.iter().filter(|player| player.connected).collect();
    ranked.sort_by(|a, b| {
        accumulated_of(b)
            .cmp(&accumulated_of(a))
            .then_with(|| a.name.cmp(&b.name))
    });

    SystemAdminSnapshot {
        players: ranked
            .into_iter()
            .enumerate()
            .map(|(index, player)| SystemPlayerEntry {
                rank: (index + 1) as i64,
                player_id: player.player_id.clone(),
                name: player.name.clone(),
                accumulated_score: accumulated_of(player),
                crossword_score: player.crossword_score,
                word_search_score: player.word_search_score,
                quiz_score: player.quiz_score,
            })
            .collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn player(player_id: &str, name: &str) -> SystemScoreInput {
        SystemScoreInput {
            player_id: player_id.to_string(),
            name: name.to_string(),
            connected: true,
            crossword_score: None,
            word_search_score: None,
            quiz_score: None,
        }
    }

    #[test]
    fn ranks_connected_players_ignoring_quiz() {
        let board = build_system_leaderboard(&[
            SystemScoreInput {
                crossword_score: Some(100),
                quiz_score: Some(5000),
                ..player("quiz-heavy", "Quiz")
            },
            SystemScoreInput {
                word_search_score: Some(1100),
                crossword_score: Some(200),
                ..player("search", "Search")
            },
            SystemScoreInput {
                connected: false,
                crossword_score: Some(9000),
                ..player("away", "Away")
            },
            player("new", "New"),
        ]);

        let names: Vec<_> = board.players.iter().map(|entry| entry.name.as_str()).collect();
        assert_eq!(names, ["Search", "Quiz", "New"]);
        let scores: Vec<_> = board.players.iter().map(|entry| entry.accumulated_score).collect();
        assert_eq!(scores, [1300, 100, 0]);
        let quiz: Vec<_> = board.players.iter().map(|entry| entry.quiz_score).collect();
        assert_eq!(quiz, [None, Some(5000), None]);
        assert_eq!(board.players[0].rank, 1);
        assert_eq!(board.players[2].rank, 3);
    }

    #[test]
    fn breaks_ties_by_name() {
        let board = build_system_leaderboard(&[
            SystemScoreInput {
                crossword_score: Some(100),
                ..player("b", "Bea")
            },
            SystemScoreInput {
                word_search_score: Some(100),
                ..player("a", "Ada")
            },
        ]);
        let names: Vec<_> = board.players.iter().map(|entry| entry.name.as_str()).collect();
        assert_eq!(names, ["Ada", "Bea"]);
        let ranks: Vec<_> = board.players.iter().map(|entry| entry.rank).collect();
        assert_eq!(ranks, [1, 2]);
    }
}

File: libs/shared/src/system.ts
/** Sent with `player:kicked` when the host removes someone from the party. */
export const SYSTEM_REMOVAL_REASON = 'Removed from the system by admin';

/** Sent with `player:kicked` when the quiz host ejects someone from the quiz only. */
export const QUIZ_REMOVAL_REASON = 'Removed from the quiz game';

export interface SystemPlayerEntry {
  rank: number;
  playerId: string;
  name: string;
  /** Crossword score + word search score. Quiz points are excluded. */
  accumulatedScore: number;
  /** Null when the player has not entered the crossword. */
  crosswordScore: number | null;
  /** Null when the player has not entered the word search. */
  wordSearchScore: number | null;
  /** Null when the player has not joined the quiz. */
  quizScore: number | null;
}

export interface SystemAdminSnapshot {
  players: SystemPlayerEntry[];
}

export interface SystemScoreInput {
  playerId: string;
  name: string;
  connected: boolean;
  crosswordScore: number | null;
  wordSearchScore: number | null;
  quizScore: number | null;
}

function accumulatedOf(player: SystemScoreInput): number {
  return (player.crosswordScore ?? 0) + (player.wordSearchScore ?? 0);
}

/** Connected players, ranked by crossword + word search score. */
export function buildSystemLeaderboard(
  players: SystemScoreInput[]
): SystemAdminSnapshot {
  const ranked = players
    .filter((player) => player.connected)
    .sort((a, b) => {
      const delta = accumulatedOf(b) - accumulatedOf(a);
      if (delta !== 0) {
        return delta;
      }
      return a.name.localeCompare(b.name);
    });

  return {
    players: ranked.map((player, index) => ({
      rank: index + 1,
      playerId: player.playerId,
      name: player.name,
      accumulatedScore: accumulatedOf(player),
      crosswordScore: player.crosswordScore,
      wordSearchScore: player.wordSearchScore,
      quizScore: player.quizScore,
    })),
  };
}
